### DISCUSSION TITLE
Programatically triggering a snippet

### DISCUSSION THREAD

**Discussion by s1n7ax** [L3MON4D3/LuaSnip#1108] [Q&A] [answered]
**Requirement:**

I think LuaSnip can be used in code refactoring. For instance `extract_function` could be implemented using LuaSnip. Following would be the steps usually.

- User selects some statements in visual mode
- User fire the keymap for `extract_function`
- A popup will appear to enter the name of the function
- User types the function and hit enter
- A new function will be created with the selected text as the function body

This is nice but LuaSnip could do it better IMO.

- User selects some statements in visual mode
- User fire the keymap for `extract_function_luasnip`

```
vim.keymap.set({ 'x', 'v' }, function ()
  require('luasnip').trigger(sn(1, fmt([[
  function {}(): {} {{
    {}
  }}
  ]], {
    ...
  })))
end, {})
``` 
- A function will be created. Neovim will be in Select mode name of the function is selected to enter the name
- LuaSnip `ls.jump()` can be used to jump to next and prev node

So I'm just wondering if something like this is possible right now or will there be something similar?

**Comment by L3MON4D3** [comment, marked as answer]
Oh, yeah certainly!! Look for `snip_expand` in DOC.md, you can expand any snippet, you just need to have the snippet-object (created by `s(...)`) (there are some other discussions on this, I'm too lazy to look for the exact ones right now, but search should turn them up :) ).
One issue with `snip_expand` is that it can't handle eg. snippets with `regTrig`/non-plaintext `trigEngine` (we don't know whether the text at the expand-position matches the pattern/regex), or more precisely, it can handle them, but it won't set the `env`-variables.

Regarding your suggestion, that's very possible!
You'd have to set `expand_params.env_override.VISUAL` to the text selected in visual mode, and then use it inside the snippet (eg. inside a functionNode like `f(function(args, parent) return parent.snippet.env.VISUAL end)`)

Regarding refactoring in general, I think the [`Treesitter-Postfix-Snippet`](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#treesitter-postfix-snippet) can come in very useful there because you have a convenient mechanism (treesitter) for capturing specific regions of the syntax-tree. Check the DOC.md-entry I linked, there's a nice example for turning a local function into a variable :D
Right now the treesitter is very coupled to the postfix-part, but it would be certainly possible to make it so that you press a key-combo over some treesitter-node, and it is extracted and its captures passed into a snippet, which could even be expanded elsewhere (though that would require a fair bit of setup :D )
